### ISSUE TITLE
Audit ModuleToKore for non-trivial operations

### ISSUE THREAD

**Issue by Baltoli** [runtimeverification/k#3444]
We currently do a lot of "extra" work inside ModuleToKore that would be better implemented as explicit compiler passes; Pyk doesn't need to do this, which indicates that ModuleToKore could be simplified down to a trivial Kast -> Text transformation.

The first step towards simplifying the code is to dig through it and find places where we're doing this extra work, and make it an explicit compiler pass _before_ ModuleToKore runs.

**Comment by ehildenb** [comment]
Examples:

- Adding attributes: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L249
- Generating assoc/comm/idem/unit axioms: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L440
- Generating Ceil axioms: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L566
- Inserting priorities: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L698
- Generating no-junk/overloads: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L711
- Throwing errors in ModuleToKore can become checks over the Kast: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L1274
- Adding Kore attributes: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L1493
- Converting side-conditions to ML predicates: https://github.com/runtimeverification/k/blob/a017998b1e2c8c8e6c84a4349cd2374005fe595b/kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java#L1641

**Comment by radumereuta** [comment]
The last few commits referenced here have been reverted since they introduced noticeable performance regressions.
Please revisit when we have a better understanding of what went wrong.
The reverts: #3554 #3560

**Comment by dwightguth** [comment]
I'm pretty sure the performance regression is related to the fact that I spotted duplicate instances of certain rules in the definition before this change was reverted and now they're gone. I would suggest double checking that you didn't pass `module.sentences()` somewhere that `module.localSentences()` was appropriate. This type of duplication often happens due to a mistake where a module is transformed to include all the sentences from its recursive imports.
